src/oas3/transformers/callbacks.ts

```diff
@@ -0,0 +1,28 @@
+import { IHttpCallbackOperation } from '@stoplight/types';
+import { CallbacksObject } from 'openapi3-ts';
+import { transformOas3Operation } from '../operation';
+
+export function translateToCallbacks(callbacks?: CallbacksObject): IHttpCallbackOperation[] | undefined {
+  if (!callbacks) return;
+
+  return Object.entries(callbacks).reduce((results: IHttpCallbackOperation[], [callbackName, path2Methods]) => {
+    for (const [path, method2Op] of Object.entries(path2Methods)) {
+      for (const [method, op] of Object.entries(method2Op as { [key: string]: {} })) {
+        results.push({
+          ...transformOas3Operation({
```

Comment:
🤔 kind of odd recursive thing going on here. Probably need to give it the complete original document instead of this shim - `transformOas3Operation` uses data like `operation.servers || pathObj.servers || document.servers` and the security schemes from the oas document, which is not included here.

Reply:
Right, this may look odd. However:

- I don't think servers make sense inside callback operation. CB op calls different server than those defined in top-level servers prop. The called server is defined inside CB path. That's why I think it is reasonable to leave servers empty in CB op.

- Security schemes also make little sense for callback op. OAS spec defines security schemes of the API itself. Callback is a bit different. It targets some different host and most of the time security is implemented via apiKey in url. See the example below:
```
CLIENT                   API
call /subscribe
                         call {$response.body#/url}&apiKey={$response.body#/apiKey}
```

So if we agree that security and servers are not applicable for callback operation let's think whether `IHttpCallbackOperation` should contain `servers` and `security` fields at all. On one hand, it makes it fully compatible with `IHttpOperation`, on the other, it makes no sense in context of the callback.

WDYT about creating IHttpCallbackOperation based on picking/omitting fields from IHttpOperation. Instead of fully inheriting (as it is now).

Reply:
@philsturgeon @XVincentX curious your opinion too!

Reply:
> It targets some different host and most of the time security is implemented via apiKey in url

Most of the time, or all of the time? ;)

Is token auth in a header not a thing? Or in general, how would one indicate that a callback requires apikey auth in the url? Pretty sure that callbacks can indicate that they use xyz security scheme(s) defined in the root oas document, similar to regular http operations, so we probably need to support that - what do you think?

Reply:
>  Probably need to give it the complete original document 

I have checked out the code of that function and in reality only 1% of the document is used and what's needed can be easily be passed as a standalone parameter instead of using the whole document. This would simplify the things for the callbacks a lot. `path, servers, security` — that's all we need. I suggest we pursue this path @StefanDywersant 

> I don't think servers make sense inside callback operation. CB op calls different server than those defined in top-level servers prop. The called server is defined inside CB path. That's why I think it is reasonable to leave servers empty in CB op.

I am on the opposite side. The spec says that the element of a Callback Object is a Path, and we should fully support it; that means we should fallback to the global servers if required. I think it's kind of regular to see callbacks to external APIs that are still under the same domain.

> Security schemes also make little sense for callback op.

Same as above, if the callback element is a Path, we should fully support it.

> I do not think I have ever seen anyone sending an API Key over the URL.

Unfortunately there are companies doing it. [Uala](https://uala.it) is one of these and I am afraid we in Stoplight are doing the same in api-gen3. We're likely leaking our apiKeys in all the logging system we're using.

Reply:
@XVincentX we should stop that, but also I have never seen it _for callbacks_. Sending an "API Key" in a URL would be a little wonky, most webhooks I've seen send a signature - which cannot be described in OpenAPI currently) or a HTTP Basic / Digest.

Reply:
> @XVincentX we should stop that, but also I have never seen it _for callbacks_. Sending an "API Key" in a URL would be a little wonky, most webhooks I've seen send a signature - which cannot be described in OpenAPI currently) or a HTTP Basic / Digest.

Callbacks give you full flexibility in how you implement security. ApiKey is an example. I believe most secure way is to generate single-use key for each callback at the moment of calling the enclosing operation. That's it, simple and secure (see here https://sendgrid.com/blog/whats-webhook/). I think Github or Stripe implemented that signing in order make folks life easier. Both, GH and Stripe treat their signing as optional, leaving security decisions to the caller.

Reply:
> I believe most secure way is to generate single-use key for each callback at the moment of calling the enclosing operation. That's it, simple and secure

Totally agree with you — this was never about what's the best practice, but more about "it's a scenario and even if we do not like it from a security standpoint, we might need to support it"

To complete the information, another kind of spread way is simply to ping back the original server that sent the web hook to make sure it is effectively coming from them. Paypal's been using this [for long time](https://developer.paypal.com/docs/api/webhooks/v1/#verify-webhook-signature)

> CB op calls different server

This is a strong assumption I wouldn't be so sure about.

---

I still kind of think we should fully support it; I take the argument about the security, but `servers` to me looks a legit use case. Also, I take the argument that this is a pain in the ass to support :)

We can probably go baby steps here and support the whole toy if people will ask it; so I propose

1. Ignore security and servers at the moment — we'll document the current limitation in Prism when the feature gets done
2. Try to see if we can still refactor this function to take only the parameters it really needs (path, security, servers) from the original document?

@marbemac @philsturgeon @StefanDywersant

Reply:
I'm good with the baby steps proposal!

Reply:
@StefanDywersant Pinging you on this just to make sure you do not lose the comment.

Reply:
Changing the signature might be problematic, `transformOas3Operation` is directly used in `prism`, `platform` and `graphite`.

A version compatible with both would look like: `transformOas3Operation({ path, method, document, operation, pathObj, globalServers, globalSecurities})`. As you can see it is super confusing of how to use it. TS has no deprecated keyword also which would help here.

I don't like the above, I don't have name proposal for alternative version of that func. I don't think it will be handy for end-users to extract pathObj/globalSecurity/globalServer on their own.

I vote no refactor 👎

Reply:
The only use case I see for `servers` is that the service defines a callback that will hit itself. I can't think about any real-world example for that. If it hits itself then why it does not do that internally? It would look like a bad design of that service.

I still consider callbacks as a `separate context` when considering `servers` and `security`.

Maybe this is a good question for OpenAPI Specification WG guys?

Reply:
> The only use case I see for servers is that the service defines a callback that will hit itself. 

I feel you're narrowing down the domain too much. There are a lot of companies with different teams with different APIs on different files on the same server. Think simply something like

`api.stoplight.io/prism`
`api.stoplight.io/platform`
`api.stoplight.io/next`
`api.stoplight.io/studio`

It's really likely they won't be on the same document and owned by the same teams and still, callbacks within the same organisation will be a thing.

Anyway — this is becoming a product decision which I do not own. @philsturgeon @marbemac this is your call; I am ok leaving it as it is right now (no server and no security support for callbacks at the moment) but if you have another idea, please say it now.

About the refactor, I would prefer to see this done because in the current state (to me) it does not look great but I understand the frustration of getting it done. I'm ok leaving as it is right now and rework this on the next occasion.

Reply:
> It's really likely they won't be on the same document and owned by the same teams and still, callbacks within the same organisation will be a thing.

I don't think there is an use-case for public-facing api. However, you convinced me that there is a use-case for documenting **internal organization APIs**.

Because of that I'd say, let's implement security and servers. 

There is quite a lot of effort to do that at prism level: faking security, constructing server (variables!). Http-spec will also require some love.

Currently, Callbacks PR is 1k LoC. Are you ok with splitting it into two PRs? (1. what we have now 2. servers and security).

> About the refactor, I would prefer to see this done because in the current state (to me) it does not look great but I understand the frustration of getting it done. I'm ok leaving as it is right now and rework this on the next occasion.

I don't see a solution that will improve the code here. E.g. your proposed solution makes that function problematic to use outside. Those lines https://github.com/stoplightio/http-spec/blob/master/src/oas3/operation.ts#L20-L28 would need to be moved to the caller space. So since the function is used by `prism`, `platform` and `graphite/studio` we would need to move those checks there.

Reply:
No rush @StefanDywersant — this is a product decision that I can't take alone; @philsturgeon is probably the folk deciding if that's critical or not for the feature to be released.

Reply:
I think we go baby steps for now, and do security and servers later once we have real world use cases to help drive the decision.

Reply:
Yeah I think baby steps (not dealing with servers or security) will be fine. People always let us know what they want. 

Right now I am unsure how servers would fit in, and security is going to be a lot of work to get right, as we currently only handle validating correctness, we do not handle generating an example request with reasonable default tokens. 

Seeing as a lot of people use signatures (which OpenAPI does not directly support) we can probably just skip it entirely.